presto-hive/src/main/java/com/facebook/presto/hive/statistics/QuickStatsProvider.java

```diff
@@ -0,0 +1,318 @@
+import org.apache.hadoop.fs.Path;
+import org.weakref.jmx.Managed;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.facebook.airlift.concurrent.Threads.daemonThreadsNamed;
+import static com.facebook.presto.hive.HivePartition.UNPARTITIONED_ID;
+import static com.facebook.presto.hive.HiveSessionProperties.isQuickStatsEnabled;
+import static com.facebook.presto.hive.HiveSessionProperties.isUseListDirectoryCache;
+import static com.facebook.presto.hive.HiveUtil.buildDirectoryContextProperties;
+import static com.facebook.presto.hive.NestedDirectoryPolicy.IGNORED;
+import static com.facebook.presto.hive.NestedDirectoryPolicy.RECURSE;
+import static com.facebook.presto.hive.metastore.PartitionStatistics.empty;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.concurrent.CompletableFuture.allOf;
+import static java.util.concurrent.CompletableFuture.supplyAsync;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+import static java.util.concurrent.ForkJoinPool.commonPool;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static java.util.stream.Collectors.toMap;
+
+public class QuickStatsProvider
+{
+    public static final Logger log = Logger.get(QuickStatsProvider.class);
+    public static final long MAX_CACHE_ENTRIES = 1_000_000L;
+    private static final ExecutorService backgroundFetchExecutor = newCachedThreadPool(daemonThreadsNamed("quick-stats-bg-fetch-%s"));
+    private final ScheduledExecutorService inProgressReaperExecutor = new ScheduledThreadPoolExecutor(1, daemonThreadsNamed("in-progress-reaper"));
+    private final HdfsEnvironment hdfsEnvironment;
+    private final DirectoryLister directoryLister;
+    private final List<QuickStatsBuilder> statsBuilderStrategies;
+    private final boolean recursiveDirWalkerEnabled;
+    private final ConcurrentHashMap<String, Instant> inProgressBuilds = new ConcurrentHashMap<>();
+    private final AtomicLong requestCount = new AtomicLong(0L);
+    private final AtomicLong succesfulResolveFromCacheCount = new AtomicLong(0L);
+    private final AtomicLong succesfulResolveFromProviderCount = new AtomicLong(0L);
+    private final Duration cacheExpiryDuration;
+    private final long reaperExpiryMillis;
+    private final Cache<String, PartitionStatistics> partitionToStatsCache;
```

Comment:
If you do something like

set quickStatsEnabled=true
query1
set quickStatsEnabled=false
query1

The stats computed for the first query are not used for the second even though available. Is this the expected behavior? Now if the second query were a query that references a table used in the first, and the flag were disabled by oversight, then it is a shame that the already computed stats were not used.

Related question - Did you consider making partitionToStatsCache part of CachingHiveMetastore? The advantage would be that QuickStats, if computed, would be available to all queries and comes with its own set of well-defined parameters and accounting.

Reply:
The `quick_stats_enabled` flag is meant to provide the user a switch to use (or compute first and then use) stats outside of HMS. If the flag is`false`, I think users would be surprised if they see stats being inferred for partitions they know are missing stats. The behavior is similar to how `use_history_based_plan_statistics` controls if HBO computed plan nodes stats are used or not

I did consider having these stats persisted to HMS (say a background thread that atomically updates the metastore if stats for the partition are missing), but decided to punt on this for now